hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/MergeNormalizer.java

```diff
@@ -0,0 +1,127 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master.normalizer;
+
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.hbase.HBaseIOException;
+import org.apache.hadoop.hbase.HRegionInfo;
+import org.apache.hadoop.hbase.TableName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of MergeNormalizer Logic in use:
+ * <ol>
+ * <li>get all regions of a given table
+ * <li>get avg size S of each region (by total size of store files reported in RegionLoad)
+ * <li>two regions R1 and its neighbour R2 are merged, if R1 + R2 &lt; S, and all such regions are
+ * returned to be merged
+ * <li>Otherwise, no action is performed
+ * </ol>
+ * <p>
+ * Region sizes are coarse and approximate on the order of megabytes. Also, empty regions (less than
+ * 1MB) are also merged if the age of region is &gt MIN_DURATION_FOR_MERGE (default 2)
+ * </p>
+ */
+
+@InterfaceAudience.Private
+public class MergeNormalizer extends BaseNormalizer {
+  private static final Logger LOG = LoggerFactory.getLogger(MergeNormalizer.class);
+  private static final int MIN_REGION_COUNT = 3;
+  private static final int MIN_DURATION_FOR_MERGE = 2;
+
+  @Override
+  public List<NormalizationPlan> computePlanForTable(TableName table) throws HBaseIOException {
+    if (!shouldNormalize(table)) {
+      return null;
+    }
+    List<NormalizationPlan> plans = new ArrayList<>();
+    List<HRegionInfo> tableRegions =
+        masterServices.getAssignmentManager().getRegionStates().getRegionsOfTable(table);
+    double avgRegionSize = getAvgRegionSize(tableRegions);
+    LOG.debug("Table {}, average region size: {}", table, avgRegionSize);
+    LOG.debug("Computing normalization plan for table: {}, number of regions: {}", table,
+      tableRegions.size());
+
+    int candidateIdx = 0;
+    while (candidateIdx < tableRegions.size()) {
+      if (candidateIdx == tableRegions.size() - 1) {
+        break;
+      }
+      HRegionInfo hri = tableRegions.get(candidateIdx);
+      long regionSize = getRegionSize(hri);
+      HRegionInfo hri2 = tableRegions.get(candidateIdx + 1);
+      long regionSize2 = getRegionSize(hri2);
+      if (regionSize >= 0 && regionSize2 >= 0 && regionSize + regionSize2 < avgRegionSize) {
+        // atleast one of the two regions should be older than MIN_REGION_DURATION days
+        if (isOldEnoughToMerge(hri) || isOldEnoughToMerge(hri2)) {
+          LOG.info(
+            "Table {}, small region size: {} plus its neighbor size: {}, less than the avg size "
+                + "{}, merging them",
+            table, regionSize, regionSize2, avgRegionSize);
+          plans.add(new MergeNormalizationPlan(hri, hri2));
+          candidateIdx++;
+        }
```

Comment:
For example, this could go to a separate, abstract method _computeNormalization_ in the base parent class, then _MergeNormalizer_ would implement it as it is now, and _SimpleRegionNormalizer_ would do it the way it does now.